remove empty mpi-serial settings from config_machines.xml

jedwards4b opened the pull request:
Remove the requirement to have mpilib mpi-serial explicitly listed in the config_machines.xml file since it is assumed supported on all systems.   Remove the empty executable field for that library.

Test suite: scripts_regression_tests.py, preview_run (thanks for this!), SMS_Mmpi-serial.f19_g17.X.hobart_intel
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes #1220 

User interface changes?: no

Code review:

jayeshkrishna commented:
mpi-serial does not work on Mira/Cetus (and we had previously decided it wasn't worth the effort to look at it - no one uses mpi-serial on Cetus/Mira)

jedwards4b commented:
I don't think that this changes the current behavior of mpi-serial on cetus/mira - do you want to add a more specific error message if someone tries to use mpi-serial there?

jayeshkrishna commented:
On Mira I think we can currently just remove mpi-serial from the list of supported mpi libs and the user would get a meaningful error message (mpi-serial is not supported...). I wanted to retain that.

jedwards4b commented:
mpi-serial is now included by default in the  list of mpilibs - we need a special case for mira

jayeshkrishna commented:
@jedwards4b : How often is mpi-serial used (in your experience)? I am not sure if the library is useful to HPC devs anymore.

jedwards4b commented:
It's used regularly for testing - mira/cetus is the only machine it doesn't work on.

jayeshkrishna commented:
So should we fix it on Mira/Cetus rather than adding a special case?

And why is it used for testing, why not build with an MPI library?

rljacob commented:
mpi-serial was meant for developers working on their laptop or workstation so they can run without installing an MPI library.  If that use case doesn't exist, we don't need to support mpi-serial.  Its certainly not used by people developing/running on HPC platforms/clusters.

rljacob commented:
That can of worms should probably be a different issue.   For this one, we need a way of saying a machine doesn't support mpi-serial.  There may be others besides mira/cetus down the line.

jedwards4b commented:
How about this?
``` 
<mpirun mpilib="mpi-serial">		
    <executable> unsupported</executable>		
   </mpirun>		
```

jayeshkrishna commented:
Another soln would be to retain mpi-serial in MPILIBS for supported machine (machines without mpi-serial in MPILIBS don't support it) and get rid of mpirun section for mpi-serial (if the user has chosen mpi-serial you don't need mpirun settings)

ekluzek commented:
Let me point out some important aspects of mpi-serial. A set of users that do a lot of work with this are CLM users's doing single point simulations to compare to a single tower site. Also mpi-serial is the simplest system you can put together, so if you are porting a cime-driven model and all you care about is running on a single processor, it's going to be easier to setup mpi-serial than an mpi driven system. It's easier to get a debugger (or profiler) to be connected to mpi-serial and I think debugging, running is going to be simpler, easier, and probably faster than with mpi for a single task. It's nice to be able to show that the system works in the simplest case possible. And as I said there is a significant community of CLM users that do single point cases. Also note that CAM has SCAM the single column model, and I noticed that ocean/ice added single column capability recently as well.

jedwards4b commented:
It makes a lot more sense to me to have an exception for the one machine that doesn't support it than a required entry for all of the machines that do.

jedwards4b commented:
I've updated the PR to write a meaningful error message if config_machines.xml indicates that
mpi-serial is not supported using the syntax
```
<mpirun mpilib="mpi-serial">		
    <executable> unsupported</executable>		
   </mpirun>		
```
and added this to cetus and mira machine definitions.

jayeshkrishna commented:
@jedwards4b : How about leaving mpi-serial in MPILIBS for supported machines like I mentioned above (if mpi-serial is not in MPILIBS it is not supported on the machine)? And remove mpirun sections for mpi-serial (mpi-serial does not require any specific runtime env).
The MPILIBS section indicates the mpi libraries supported on a specific machine and specifying the support for mpi-serial there would make the intent clear.

rljacob commented:
Just leave things as they are.  It does what we need.

jgfouca commented:
I read the conversation chain. For what it's worth, I like @jedwards4b solution of explicitly adding unsupported for the couple machines that can't handle mpi-serial.

rljacob commented:
I went back and read the issue this is addressing and this comment https://github.com/ESMCI/cime/issues/1220#issuecomment-285175015 confused me because it said the requirement to list mpi-serial had already been removed.

jedwards4b commented:
@rljacob The problem was that although the requirement to list mpi-serial had been removed the block 
```
<mpirun mpilib="mpi-serial">		
    <executable></executable>		
   </mpirun>	
```
was still required.   This PR removes that requirement as well.

rljacob commented:
Ok can you do that while leaving the listing of mpi-serial in MPILIBS alone?

jedwards4b commented:
If you want to have the mpi-serial listed explicitly in acme config_machines.xml this PR does not disallow that.

jedwards4b commented:
I've added the mpi-serial back to the list of mpilibs in the acme config_machines.xml

rljacob commented:
If its not listed, and a user tries to build with it, will they still get an error message?

jedwards4b commented:
That error message is provided with the additional 
```
<mpirun mpilib="mpi-serial">		
    <executable> unsupported</executable>		
   </mpirun>		
```

rljacob commented:
All the issue asked to do is remove the empty mpirun blocks for mpi-serial.  Can't you do that and leave the rest alone?

jedwards4b commented:
No - for two reasons 
1.  It doesn't do what you seem to think it does now
2. It's not what we want on the cesm side. 

I have now updated to give you the functionality you seem to think was already there and to give us the functionality that we want.

rljacob commented:
No that's a bad idea too.  We need fewer acme/cesm special cases in the core library, not more.

jedwards4b commented:
So stop trying to create a special case for acme.

rljacob commented:
Fair enough.   I'm trying to find a solution that works for everybody.   Making mpi-serial in to an assumed available library has other implications that we need to think through.

jedwards4b commented:
But it's been that way for quite some time.

rljacob commented:
After talking on the CIME telecon, we decided to go with the original PR.  The error message will come from the failed build of mpi-serial.  We don't want those empty mpirun blocks in the xml.  We don't want to introduce more low-level diffs between acme and cesm.

jedwards4b commented:
Okay - I think that this is ready for a final review and merge.
